File: day04p1/main.py
#!/usr/bin/env python3

import sys
from collections import Counter

fname = sys.argv[1] if len(sys.argv[1:]) == 1 else "input.txt"

data = open(fname).read().splitlines()
height = len(data)
width = len(data[0])

xmas = "XMAS"

total = 0

for word in [xmas, xmas[::-1]]:
    for i, horizontal in enumerate(data):
        tmp = horizontal.count(word)
        total += tmp

    verticals = ["".join([row[col] for row in data]) for col in range(len(data[0]))]

    for vertical in verticals:
        total += vertical.count(word)
        pass

    # Thanks copilot
    def get_diagonals(matrix):
        if not matrix or not matrix[0]:
            return []

        rows, cols = len(matrix), len(matrix[0])
        diagonals = []

        # Top-left to bottom-right diagonals
        for d in range(rows + cols - 1):
            diagonal = []
            for row in range(max(0, d - cols + 1), min(rows, d + 1)):
                col = d - row
                diagonal.append(matrix[row][col])
            diagonals.append(diagonal)

        # Top-right to bottom-left diagonals
        for d in range(rows + cols - 1):
            diagonal = []
            for row in range(max(0, d - cols + 1), min(rows, d + 1)):
                col = cols - 1 - (d - row)
                diagonal.append(matrix[row][col])
            diagonals.append(diagonal)

        return ["".join(d) for d in diagonals]

    for diagonal in get_diagonals(data):
        total += diagonal.count(word)
        pass


print(total)

File: day04p1/main.ts
function occurrences(bigString: string, subString: string): number {
  return (bigString.match(new RegExp(subString, "g")) || []).length;
}

export function xmasSearch(data: string[]): number {
  const xmas = "XMAS";
  const height = data.length;
  const width = data[0].length;
  let ret = 0;

  const searchPatterns = [xmas, xmas.split("").reverse().join("")];

  const searchInDirection = (getChars: (i: number, j: number) => string) => {
    for (const searchWord of searchPatterns) {
      for (let i = 0; i < height; i++) {
        let chars = "";
        for (let j = 0; j < width; j++) {
          chars += getChars(i, j);
        }
        ret += occurrences(chars, searchWord);
      }
    }
  };

  // Horizontal and vertical searches
  searchInDirection((i, j) => data[i][j]); // Horizontal
  searchInDirection((i, j) => data[j][i]); // Vertical

  // Diagonal searches
  const searchDiagonals = (startRow: number, startCol: number, rowInc: number, colInc: number) => {
    for (let i = 0; i < height; i++) {
      let chars = "";
      for (let j = 0; j < width; j++) {
        const row = startRow + i * rowInc;
        const col = startCol + j * colInc;
        if (row >= 0 && row < height && col >= 0 && col < width) {
          chars += data[row][col];
        }
      }
      ret += occurrences(chars, xmas);
      ret += occurrences(chars, xmas.split("").reverse().join(""));
    }
  };

  // Diagonal (\) and (/)
  searchDiagonals(0, 0, 1, 1); // Diagonal (\) from top-left
  searchDiagonals(height - 1, 0, -1, 1); // Diagonal (/) from bottom-left

  return ret;
}

if (import.meta.main) {
  const bytes = await Deno.readFile("input.txt");
  const fileStr = new TextDecoder().decode(bytes).trimEnd();
  const lines = fileStr.split("\n");
  console.log(xmasSearch(lines));
}